src/throttle_module.rs

```diff
@@ -83,57 +84,76 @@ impl UnpreparedThrottleModule {
 }
 
 impl ThrottleModule {
-    fn disable_control(&mut self, debug_console: &mut DebugConsole) {
+    pub fn disable_control(&mut self, debug_console: &mut DebugConsole) -> Result<(), OxccError> {
+        let mut result = Ok(());
+
         if self.control_state.enabled {
             self.accelerator_position.prevent_signal_discontinuity();
 
-            self.throttle_dac.output_ab(
+            if let Err(_) = self.throttle_dac.output_ab(
                 DacOutput::clamp(self.accelerator_position.low()),
                 DacOutput::clamp(self.accelerator_position.high()),
-            );
+            ) {
+                result = Err(OxccError::IO);
+            }
 
+            // even if we've encountered an error, we can still disable
             self.throttle_pins.spoof_enable.set_low();
             self.control_state.enabled = false;
             writeln!(debug_console, "Throttle control disabled");
         }
+
+        result
     }
 
-    fn enable_control(&mut self, debug_console: &mut DebugConsole) {
+    fn enable_control(&mut self, debug_console: &mut DebugConsole) -> Result<(), OxccError> {
+        let mut result = Ok(());
+
         if !self.control_state.enabled && !self.control_state.operator_override {
             self.accelerator_position.prevent_signal_discontinuity();
 
-            self.throttle_dac.output_ab(
+            if let Err(_) = self.throttle_dac.output_ab(
                 DacOutput::clamp(self.accelerator_position.low()),
                 DacOutput::clamp(self.accelerator_position.high()),
-            );
-
-            self.throttle_pins.spoof_enable.set_high();
-            self.control_state.enabled = true;
-            writeln!(debug_console, "Throttle control enabled");
+            ) {
+                result = Err(OxccError::IO);
+            } else {
+                self.throttle_pins.spoof_enable.set_high();
+                self.control_state.enabled = true;
+                writeln!(debug_console, "Throttle control enabled");
+            }
         }
+
+        result
     }
 
-    fn update_throttle(&mut self, spoof_command_high: u16, spoof_command_low: u16) {
+    fn update_throttle(
+        &mut self,
+        spoof_command_high: u16,
+        spoof_command_low: u16,
+    ) -> Result<(), OxccError> {
         if self.control_state.enabled {
             // TODO - revisit this, enforce high->A, low->B
             self.throttle_dac.output_ab(
                 ranges::coerce(ThrottleSpoofHighSignal::clamp(spoof_command_high)),
                 ranges::coerce(ThrottleSpoofHighSignal::clamp(spoof_command_low)),
-            );
+            )?;
```

Comment:
If this is an entirely private and internal function, you may consider *not* converting the error type to the top-level/global error type immediately, and instead return the more immediate/domain-specific error type so that the other wrapping functions (still in this module) can make the choice about how to handle it without having to destructure/dig back down to the right variant or otherwise deal with dropped/simplified error state.

Reply:
Great idea, I'll work on moving fault handling mechanics into the modules :+1: